File: typescript/src/functors/builtin.ts
/**
 * Basic builtin functors for JSE.
 *
 * Following docs/regular.md basic operators section:
 * - $quote: Quote unevaluated expression
 * - $eq: Equality comparison
 * - $cond: Multi-branch conditional
 * - $head/$tail: List car/cdr operations
 * - $atom?: Check if value is JSON atom type
 * - $cons: Construct list (Clojure-style)
 */

import type { JseValue } from "../types.js";
import type { Env } from "../env.js";

// Type alias for functors
export type Functor = (env: Env, ...args: JseValue[]) => JseValue;

/**
 * Return argument without evaluation.
 */
export function _quote(env: Env, ...args: JseValue[]): JseValue {
  return args[0] ?? null;
}

/**
 * Compare two values for equality.
 */
export function _eq(env: Env, ...args: JseValue[]): JseValue {
  if (args.length !== 2) {
    throw new Error("$eq requires exactly 2 arguments");
  }
  return args[0] === args[1];
}

/**
 * Multi-branch conditional evaluation.
 * Forms:
 * - [$cond, test1, result1, test2, result2, ...]
 * - [$cond, test1, result1, test2, result2, ..., default]
 */
export function _cond(env: Env, ...args: JseValue[]): JseValue {
  if (args.length === 0) {
    return null;
  }

  let pairs: [JseValue, JseValue][];
  let defaultCase: JseValue | null = null;

  if (args.length % 2 === 0) {
    // Pairs only: [test1, result1, test2, result2, ...]
    pairs = [];
    for (let i = 0; i < args.length; i += 2) {
      pairs.push([args[i], args[i + 1]]);
    }
  } else {
    // With default: [test1, result1, ..., default]
    pairs = [];
    for (let i = 0; i < args.length - 1; i += 2) {
      pairs.push([args[i], args[i + 1]]);
    }
    defaultCase = args[args.length - 1];
  }

  for (const [test, result] of pairs) {
    if (env.eval(test)) {
      return env.eval(result);
    }
  }

  return defaultCase !== null ? env.eval(defaultCase) : null;
}

/**
 * Get first element of list (car).
 */
export function _head(env: Env, ...args: JseValue[]): JseValue {
  if (args.length === 0) {
    throw new Error("$head requires a list argument");
  }

  const lst = env.eval(args[0]);
  if (!Array.isArray(lst)) {
    throw new Error("$head requires a list argument");
  }
  if (lst.length === 0) {
    throw new Error("$head requires non-empty list");
  }

  return lst[0];
}

/**
 * Get rest of list (cdr).
 */
export function _tail(env: Env, ...args: JseValue[]): JseValue {
  if (args.length === 0) {
    throw new Error("$tail requires a list argument");
  }

  const lst = env.eval(args[0]);
  if (!Array.isArray(lst)) {
    throw new Error("$tail requires a list argument");
  }

  return lst.length > 1 ? lst.slice(1) : [];
}

/**
 * Check if value is a JSON atom type.
 * Atoms are: number, string, boolean, null
 */
export function _atomp(env: Env, ...args: JseValue[]): JseValue {
  if (args.length === 0) {
    throw new Error("$atom? requires an argument");
  }

  const value = env.eval(args[0]);
  return (
    value === null ||
    typeof value === "number" ||
    typeof value === "boolean" ||
    typeof value === "string"
  );
}

/**
 * Conjoin element and list (Clojure-style).
 * First argument: any element
 * Second argument: must be a list
 * Returns: new list [element, ...list]
 */
export function _cons(env: Env, ...args: JseValue[]): JseValue {
  if (args.length !== 2) {
    throw new Error("$cons requires exactly 2 arguments");
  }

  const elem = env.eval(args[0]);
  const lst = env.eval(args[1]);

  if (!Array.isArray(lst)) {
    throw new Error("$cons second argument must be a list");
  }

  return [elem, ...lst];
}

// Dict of all builtin functors for registration
export const BUILTIN_FUNCTORS: Record<string, Functor> = {
  $quote: _quote,
  $eq: _eq,
  $cond: _cond,
  $head: _head,
  $tail: _tail,
  "$atom?": _atomp,
  $cons: _cons,
};

File: rust/src/functors/builtin.rs
//! Basic builtin functors for JSE.
//!
//! Following Python's functors/builtin.py pattern:

use std::rc::Rc;
use std::cell::RefCell;
use serde_json::Value;
use crate::env::{Env, Functor};
use crate::ast::AstError;

pub fn quote(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    Ok(args.first().cloned().unwrap_or(Value::Null))
}

pub fn eq(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() < 2 {
        return Ok(Value::Bool(true));
    }
    if args.len() == 2 {
        return Ok(Value::Bool(args[0] == args[1]));
    }
    // Check all adjacent pairs are equal
    for i in 0..args.len() - 1 {
        if args[i] != args[i + 1] {
            return Ok(Value::Bool(false));
        }
    }
    Ok(Value::Bool(true))
}

pub fn cond(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    for (i, arg) in args.iter().enumerate() {
        match arg {
            Value::Array(clause) => {
                if clause.is_empty() {
                    continue;
                }

                // If this is the last clause and has only one element, it's the else clause
                if i == args.len() - 1 && clause.len() == 1 {
                    return Ok(clause[0].clone());
                }

                if clause.len() < 2 {
                    return Err(AstError::EvalError("$cond clauses must have at least 2 elements except the last".to_string()));
                }

                if is_truthy(&clause[0]) {
                    return Ok(clause[1].clone());
                }
            }
            _ => {
                return Err(AstError::EvalError("$cond clauses must be arrays".to_string()));
            }
        }
    }

    Ok(Value::Null)
}

pub fn head(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.is_empty() {
        return Err(AstError::ArityError("$head requires a list argument".to_string()));
    }

    match &args[0] {
        Value::Array(arr) if !arr.is_empty() => Ok(arr[0].clone()),
        Value::Array(_) => Err(AstError::EvalError("$head requires non-empty list".to_string())),
        _ => Err(AstError::TypeError("$head requires a list argument".to_string())),
    }
}

pub fn tail(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.is_empty() {
        return Err(AstError::ArityError("$tail requires a list argument".to_string()));
    }

    match &args[0] {
        Value::Array(arr) if arr.len() > 1 => Ok(Value::Array(arr[1..].to_vec())),
        Value::Array(_) => Ok(Value::Array(vec![])),
        _ => Err(AstError::TypeError("$tail requires a list argument".to_string())),
    }
}

pub fn atomp(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.is_empty() {
        return Ok(Value::Bool(false));
    }

    let is_atom = matches!(args[0], Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_));
    Ok(Value::Bool(is_atom))
}

pub fn cons(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() != 2 {
        return Err(AstError::ArityError("$cons requires exactly 2 arguments".to_string()));
    }

    match &args[1] {
        Value::Array(arr) => {
            let mut result = vec![args[0].clone()];
            result.extend(arr.clone());
            Ok(Value::Array(result))
        }
        _ => Err(AstError::TypeError("$cons second argument must be a list".to_string())),
    }
}

fn is_truthy(v: &Value) -> bool {
    match v {
        Value::Bool(b) => *b,
        Value::Null => false,
        _ => true,
    }
}

pub fn builtin_functors() -> std::collections::HashMap<&'static str, Functor> {
    let mut m = std::collections::HashMap::new();
    m.insert("$quote", quote as Functor);
    m.insert("$eq", eq as Functor);
    m.insert("$cond", cond as Functor);
    m.insert("$head", head as Functor);
    m.insert("$tail", tail as Functor);
    m.insert("$atom?", atomp as Functor);
    m.insert("$cons", cons as Functor);
    m
}
